Fix for #613 - Edge get deleted even though it is still mapped

Andy2003 opened the pull request:
## Description

When invoking `session.save` the `EntityGraphMapper` clears potentially entities registered in the `mappingContext` in Line:

https://github.com/neo4j/neo4j-ogm/blob/9a9d458afce0bf44fbbc5d3783100329fac52336/core/src/main/java/org/neo4j/ogm/context/EntityGraphMapper.java#L199-L200

But the `mappingContext` still contains relationships related to the cleared entities. This fix removes the stale relationships in the `mappingContext`.

## Related Issue

#613 

## How Has This Been Tested?

see attached test case

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

michael-simons commented:
Thanks, we'll have a look.

Andy2003 commented:
@michael-simons / @meistermeier I don't think removing entities from the `mappingContext` is not a good idea, as this will cause edges that should not be deleted to be removed when subsequent calls to save are made.

https://github.com/neo4j/neo4j-ogm/blob/9a9d458afce0bf44fbbc5d3783100329fac52336/core/src/main/java/org/neo4j/ogm/context/EntityGraphMapper.java#L199-L200

Even with my fix above I can create additional tests which will fail.

michael-simons commented:
Sorry, but to many "nots" in your sentence :D 

So you think removing is a good idea, right?

Andy2003 commented:
I updated my comment, sorry for this. Removing the lines noted above, will fix all of my tests, but other tests will then fail.

Andy2003 commented:
The goal of both calls is to remove obsolete objects from the session. The problem with this is that the relations are still kept in the `mappingContext`. This causes the `EntityGraphMapper` to "think" that the on the second run, the newly loaded object, which is less initialized than before, has these relations deleted.

Andy2003 commented:
The solution could be to update existing entities in the `mappingContext` to reflect the changes. For example, the other side of a bidirectional mapping could also be changed by the session (see also https://github.com/neo4j/neo4j-ogm/issues/591). This means that the session no longer contains obsolete information. However, this also means that the user is aware of this behavior, since the `EntityGraphMapper` then also updates the objects that were not touched by the user. However, the user can easily overcome this circumstance by calling `session.detachNodeEntity` or `session.detachRelationshipEntity`.

Andy2003 commented:
@michael-simons I now implemented the idea of my last comment, can you please have a look at the changes and give me some feedback

michael-simons commented:
Hi Andy.
Thanks for providing the PR.
We will not include it at the moment for several reasons:
* It introduces a lot of additional complexity in an already complex area
* It doesn't seem to fix #613 completely.
* Other people having exact the opposite problem with relationships.

We will however keep 613 open.

It would be helpful discussing those suggested changes upfront in the future. You can reach out via email or the community site anytime.

Andy2003 commented:
@michael-simons 

1. why doesn't the PR seem to fix #613 completely? The tests are all green with the fix.
2. Can you please provide more details on `Other people having exact the opposite problem with relationships`? What is the opposite problem?
3. How do we deal with the bug #613 now, it is not fixed after all. Do you have an alternative solution?

michael-simons commented:
1. Even with my fix above I can create additional tests which will fail. 
2. Relationships not being deleted.
3. It is unfixed for the time being.

May a mapping that doesn't rely on circular mappings would help you? Given I interpret that test correctly

```
Node l1 = new Node("l1").setNodeType(nodeType);
        Node l2 = new Node("l2")
            .setNodeType(nodeType)
            .setLabels(Collections.singleton(new Label().setKey("label1")))
            .setChildOf(l1);

        l1.setChildNodes(Collections.singleton(l2));
```

I non-circular mapping N-<M where n has many ms can be loaded and saved without losing anything.

I value your contribution very much but I right now I cannot bring in code that changes so much and making it our own without completly understanding it. 
As I didn't have time to work on this right now, I decided it's the best compromise to keep the issue but close the PR for now. I understand that you possibly disagree, but I cannot change that.

Andy2003 commented:
Regarding point 1:
This is true for the 1st commit, with which I opened the PR, I then also created a test case that replicates this bug. But then I suggested an alternative solution (https://github.com/neo4j/neo4j-ogm/pull/615#issuecomment-470493355), and then implemented it (https://github.com/neo4j/neo4j-ogm/pull/615/commits/92267fd944f2efd4fdac5bbe52e654650da3ef56) and informed you about this change (https://github.com/neo4j/neo4j-ogm/pull/615#issuecomment-470948576)

Regarding point 2:

- my recent commit also solved this issue
- the changes I made are configurable and switched off by default

Many changes in PR are code-cleanup (like defining generics or using instance variables instead of passing them as parameters). The domain-oriented changes are fairly limited.

If you currently have no time to review the PR, I understand that. However, I wouldn't completely discard the changes and would like to reopen PR. 

Since my changes are in use in a custom build, we have no more problems with randomly deleted edges (which was the case before).

michael-simons commented:
Don't throw your changes away, again, I value them.

But: "Many changes in PR are code-cleanup (like defining generics or using instance variables instead of passing them as parameters)."

I beg to differ here. While generics most often don't break obvious stuff, changing parameters does. Spring Data Neo4j uses some internal apis and this might break. Did you test this?
We need to test this.
I'd much prefer having those changes separately.

Andy2003 commented:
@michael-simons yes we use spring-data-neo4j

michael-simons commented:
I just picked the test now and worked my way through it: You can avoid the behaviour by

* flushing the session before storing the parent
* fully initializing the parent
* not mapping both directions of the relationship

Andy2003 commented:
Hello Michal,

I updated the branch again and removed all changes that had nothing to do with the PR (code clean-ups). Unfortunately, this is no longer shown here, since you have already closed the PR.

Regarding your remarks: We use the ogm driver with spring-data-neo4j and spring-data-repositories. Your suggestions are unfortunately not really practical in our project.

> flushing the session before storing the parent

We don't want to pollute our businiess-logic with infrastructure code

> fully initializing the parent

At some point we would then have to load the entire database into memory, which is also not practicable

> not mapping both directions of the relationship

This seems to me to be the completely wrong approach for a driver, not to recommend bi-directional mapping.

Since my changes are completely disabled by default, I would appreciate it if you could open the PR again and take a closer look at it as soon you will find some time.

michael-simons commented:
> Since my changes are completely disabled by default, I would appreciate it if you could open the PR again and take a closer look at it as soon you will find some time.

Could you point me to the code where disable and enable them?

Andy2003 commented:
https://github.com/kiwigrid/neo4j-ogm/blob/aeb83c97bb8ab83b3e934e54cb9dfa9bec3b8393/core/src/main/java/org/neo4j/ogm/session/SessionFactory.java#L207-L213
